File: class4/exercises/cloneGraph.py
from collections import deque
"""
Definition for a undirected graph node
class UndirectedGraphNode:
    def __init__(self, x):
        self.label = x
        self.neighbors = []
"""
class UndirectedGraphNode:
    def __init__(self, x):
        self.label = x
        self.neighbors = []

class Solution:
    """
    @param: node: A undirected graph node
    @return: A undirected graph node
    """
    def cloneGraph(self, node):
        # write your code here
        start = node 
        
        if node is None:
            return node 
            
        nodes = self.getNodes(node)
        
        mapping = {}
        for node in nodes:
            mapping[node] = UndirectedGraphNode(node.label)
            
        for node in nodes:
            new_node = mapping[node]
            
            for neighbor in node.neighbors:
                new_neighbor = mapping[neighbor]
                new_node.neighbors.append(new_neighbor)
                
        return mapping[start]
        
    def getNodes(self, node):
        
        
        queue = deque()
        seen = set() 
        # results = set()
        
        queue.append(node)
        seen.add(node)
        # results.add(node)
        
        while queue:
            
            size = len(queue)
            
            for _ in range(size):
                
                cur_node = queue.popleft() 
                
                for neighbor in cur_node.neighbors:
                    
                    if neighbor in seen:
                        continue 
                    
                    queue.append(neighbor)
                    seen.add(neighbor)
                    # results.add(neighbor)
                    
        # return results
        return seen
        
        
                    
                    
        
        

File: interviews/Perficient/CloneGraph.java
import java.util.*; 

/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;

    public Node() {}

    public Node(int _val,List<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
class Node {
    public int val;
    public List<Node> neighbors;

    public Node() {}

    public Node(int _val,List<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};

public class CloneGraph {
    public Node cloneGraph(Node node) {
        
        List<Node> nodes = this.getNodes(node);
        
        Map<Node, Node> map = new HashMap<Node, Node>(); 
        
        for (Node curt : nodes) {
            
            map.put(curt, new Node(curt.val, new ArrayList<Node>()));
        }
        
        for (Node curt : nodes) {
            
            Node newNode = map.get(curt); 
            
            for (Node neighbor : curt.neighbors) {
                
                Node newNeighbor = map.get(neighbor);
                newNode.neighbors.add(newNeighbor);
                
            }
        }
        
        return map.get(node);
    }
    
    private List<Node> getNodes(Node node) {
        
        Queue<Node> queue = new LinkedList<Node>(); 
        Set<Node> seen = new HashSet<Node>(); 
        List<Node> nodes = new ArrayList<Node>();
        
        queue.offer(node);
        seen.add(node);
        
        while (!queue.isEmpty()) {
            
            int size = queue.size(); 
            
            for (int i = 0; i < size; i++) {
                
                Node curt = queue.poll();
                nodes.add(curt);
                
                for (Node neighbor : curt.neighbors) {
                    
                    if (seen.contains(neighbor)) {
                        continue;
                    }
                    
                    queue.offer(neighbor);
                    seen.add(neighbor);
                    
                }
            }
        }
        
        return nodes;
        
    }
}